src/core/stack_switching/calculate_wasm_func_nargs_fallback.mjs

```diff
@@ -0,0 +1,52 @@
+const checkerModules = [undefined, undefined, undefined, undefined];
+
+function makeCheckerModule(n) {
+  // Make a webassembly module of the form:
+  //
+  // (module
+  //     (import "e" "f" (func (param i32 ... i32) (result i32)))
+  // )
+  //
+  // with n i32's. We'll try to import the function to this module to see if it
+  // has this signature.
+  return new WebAssembly.Module(
+    // prettier-ignore
+    new Uint8Array([
+      0, 97, 115, 109, // magic number
+      1,  0,   0,   0, // wasm version
+      // Type section code and byte length
+      1, 5 + n,
+      1, // One type
+      // Type is (i32 i32 ... i32) => i32 where there are n i32's
+      96, n, ...Array.from({ length: n }, () => 127),
+      1, 127,
+      // Import section code and byte length
+      2, 7,
+      // One import
+      1,
+      // "e" "f"
+      1, 101, 1, 102,
+      //   A function of type 0
+      0, 0,
+    ]),
+  );
+}
+
+function getCheckerModule(n) {
+  if (checkerModules[n] === undefined) {
+    checkerModules[n] = makeCheckerModule(n);
+  }
+  return checkerModules[n];
+}
+
+export function calculateWasmFuncNargsFallback(functionPtr) {
+  for (let n = 0; n < 4; n++) {
```

Comment:
How is the number 4 selected?

Reply:
Yeah this also needs a comment.

It matches up with the cases here:
https://github.com/pyodide/pyodide/blob/main/cpython/patches/0004-Make-Emscripten-trampolines-work-with-JSPI.patch?plain=1#L290-L309

Basically, we're trying to adapt for all bad fpcast cases. The most common case where we need the trampoline is that the function is a `METH_NOARGS` function which either takes 2 arguments (the correct number) or 1 argument (dropping the `NULL` argument) or 0 arguments (dropping both `NULL` and the `module` argument). Or people could do a `METH_VARARGS | METH_KWDS` which should have signature `PyObject* mod, PyObject* args, PyObject* kwds` and ignore either the kwds, both the kwds and the args, or everything, so 3, 2, 1, or 0 arguments.

We're importing the function with various signatures `(n i32's) => i32` to find out which case it's in. If it's not one of 0, 1, 2, or 3 i32 arguments, then we're in some other weird case. For instance it might return a `f64` or something. The type reflection code wouldn't notice this and would attempt the call_indirect raising a fatal indirect call signature mismatch. This code would notice and raise a nonfatal `SystemError`.

Reply:
Got it. Thank for the explanation. Could you add the comment to the code as well?